### PR TITLE
LPS-186911 Create the customEmptyState prop on FDS

### PR DESCRIPTION
**Pull request by rebsilva**
Related Issue: https://issues.liferay.com/browse/LPS-186911

Hi reviewer!

This PR has the objective to make possible the use of custom empty states on FDS; the necessity came from [this bug](https://issues.liferay.com/browse/LPS-186168) opened by the design team. You can also see the mockups with the expected messages [here](https://www.figma.com/file/4K3ROCpdTnKAFNTpwZNZcD/80-LPS-objects-empty-states-investigation?type=design&node-id=502-95200&t=L4AMjXenpk1UzJVl-0).

:warning: Before forwarding this PR I will ask the Frontend Infra team to review the commit where I change the prop from the FDS, but I will wait for our team's review first. Thanks! :cherry_blossom: 

What this PR does?

- Creates a new prop called `customEmptyState` inside the FDS; if it exists, it will be rendered instead of the default
- Creates a new React component called `CustomEmptyState` that will be used as the base for all the custom empty states
- Applies the custom empty state to the table on the Predefined Values

Screenshot from 2023-06-13 15-29-37

### CODE REVIEW

**Comment by rebsilva** [comment]
The failing tests don't seem related to the PR

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx
Comment (cgmonte) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx:218]: This returns a function that returns a function. It looks like a use case for `useCallback`. Check [React docs](https://react.dev/reference/react/useMemo#memoizing-a-function).

This would look like as something like this:
```
	const getSelectedFields = useCallback(() => {
		const objectFields: ObjectField[] = [];

		predefinedValues?.forEach(({name}) => {
			if (objectFieldsMap.has(name)) {
				const field = objectFieldsMap.get(name);
				objectFields.push(field as ObjectField);
			}
		});

		return objectFields;
	}, [predefinedValues, objectFieldsMap]);
```

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx
Comment (cgmonte) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx:235]: Same as before: This returns a function that returns a function. It looks like a use case for `useCallback`. You could return the nested function directly. Check [React docs](https://react.dev/reference/react/useMemo#memoizing-a-function).

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx
Comment (cgmonte) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx:297]: Same as before: `useCallback` can avoid the function nesting. Have you tried using it?

**Review by cgmonte — COMMENTED** [review]
Hey @rebsilva Looks great to me! I just made some comments suggesting replacing `useMemo` with `useCallback` in some of the uses.

**Comment by cgmonte** [comment]
@rebsilva While testing, I reached a state where there were no fields in the object, so when I tried to click "add field" in the empty state, it did nothing. In Action -> then -> Update an object entry:

Untitled_ Jun 14, 2023 3_49 PM

This may confuse users, because the action has no feedback at all. We should inform the user why the action does nothing. Or maybe disable the Add Field button and explain why it is disabled? It is also confusing, because for the Add Field button to work, you need to... Add a Field :P But add it to the object, in another place! So renaming this can be a good a idea, maybe Add Field Action... I don't know. I think this should be discussed with the design team, if it wasn't already.

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx
Comment (rebsilva) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx:218]: Oh great! I will take a look at that, thanks Carlos!

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx
Comment (rebsilva) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx:297]: Never tried using this before, I'll take a look for sure!

**Comment by rebsilva** [comment]
Hi @cgmonte! Thanks for all your feedback :cherry_blossom:

About the last one I've never realized this behavior, but from what I understood it was already happening this with the old empty state right? Could you please list the steps you had to do to reach this scenario?

I agree we should talk with the design team about this, but in my point of view, it doesn't block this PR because the main goal here is to make custom empty states available in FDS. Let me know if you agree or not, and thanks again for catching this! :blush:

**Comment by rebsilva** [comment]
I've made the changes you mentinoed @cgmonte, could you take a look again please? Thanks!

**Comment by cgmonte** [comment]
> Hi @cgmonte! Thanks for all your feedback cherry_blossom
> 
> About the last one I've never realized this behavior, but from what I understood it was already happening this with the old empty state right? I've tested here and it happens the same on the main branch.
> 
> I agree we should talk with the design team about this, but in my point of view, it doesn't block this PR because the main goal here is to make custom empty states available in FDS. Let me know if you agree or not, and thanks again for catching this! blush

Yeah, totally agree, this does not block this PR. :relaxed:

**Review by cgmonte — APPROVED** [review]
Hey @rebsilva, if you have tested and it works, LGTM!!

**Review by naraandrade — CHANGES_REQUESTED** [review]
@rebsilva Just a small change :)

File: modules/apps/object/object-js-components-web/src/main/resources/META-INF/resources/components/CustomEmptyState.tsx
Comment (rebsilva) [reply modules/apps/object/object-js-components-web/src/main/resources/META-INF/resources/components/CustomEmptyState.tsx:20]: Oh cool! I will change it and also keep in mind for other PRs in the future, thanks Nara! :cherry_blossom:

**Review by naraandrade — APPROVED** [review]

**Comment by rebsilva** [comment]
Sent to frontend-infra team to review https://github.com/liferay-frontend/liferay-portal/pull/3457

**Pull request closed without merging**
